aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp

```diff
@@ -0,0 +1,54 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/Config.h>
+
+#include "MKLDNNHandle.h"
+
+namespace at { namespace native {
+
+#if AT_MKLDNN_ENABLED()
+
+ideep::tensor& IdeepTensorFromMKLDNNTensor(const Tensor& self) {
+  return ((MKLDNNHandle*)self.unsafeGetTensorImpl()->unsafe_opaque_handle())->get_ideep_tensor();
+}
+
+Tensor mkldnn_to_dense(const Tensor& mkldnn_tensor) {
+  // TODO: share buffer without copy when MKLDNN tensor is in plain format
+  AT_ASSERT(mkldnn_tensor.type_id() == MkldnnCPUTensorId());
+  ideep::tensor& stensor = IdeepTensorFromMKLDNNTensor(mkldnn_tensor);
+  auto dims = stensor.get_dims();
+  // NOTE: int32_t dims from ideep::tensor but sizes needs int64_t
+  Tensor cpu_tensor = at::zeros(
+    std::vector<int64_t>(dims.begin(), dims.end()),
+    mkldnn_tensor.options().layout(c10::kStrided));
+  stensor.reorder_to(cpu_tensor.template data<float>());
+  return cpu_tensor;
+}
+
+Tensor dense_to_mkldnn(const Tensor& cpu_tensor) {
+  // TODO: share CPU storage without explicit buffer copy here
```

Comment:
> how we can share memory if we do reordering here?

@dzhulgakov For `dense_to_mkldnn`, the conversion is actually not applied since we do not know what blocked format the following op requires. So we just turn the tensor type into MKL-DNN while still use the plain format (public format supported by framework, e.g. NCHW) of the CPU tensor. For `mkldnn_to_dense`, the conversion might or might not happen depending on whether MKL-DNN tensor is in plain format or not. In the case of plain format, we can still share buffer without explicit copy.

Reply:
@dzhulgakov Hmm, I was actually wrong. We cannot share memory here even if the layout is same for CPU tensor and MKL-DNN tensor since user might change the CPU tensor later. To support buffer sharing, we need to add an in-place API `to_mkldnn_` instead. Same thing for `mkldnn_to_dense` but I didn't find in-place `to_dense_` call. Can I add one?